src/proxy-animation.js

```diff
@@ -133,6 +364,175 @@ function hasActiveTimeline(details) {
+
+  // Resume of a paused animation after a timeline change snaps to the scroll
+  // position.
+  if (details.resetCurrentTimeOnResume) {
+    previousCurrentTime = null;
+    details.resetCurrentTimeOnResume = false;
+  }
+
+  const playbackRate = effectivePlaybackRate(details);
+  const upperBound = effectEnd(details);
+  if (playbackRate > 0 && autoRewind && (previousCurrentTime == null ||
+                                         previousCurrentTime < 0 ||
+                                         previousCurrentTime >= upperBound)) {
+    seekTime = 0;
+  } else if (playbackRate < 0 && autoRewind &&
+             (previousCurrentTime == null || previousCurrentTime <= 0 ||
+             previousCurrentTime > upperBound)) {
+    if (upperBound == Infinity) {
+      // Defer to native implementation to handle throwing the exception.
+      details.animation.play();
+      return;
+    }
+    seekTime = upperBound;
+  } else if (playbackRate == 0 && previousCurrentTime == null) {
+    seekTime = 0;
+  }
+
+  // 6. If seek time is resolved,
+  //        6a1. Set animation's start time to seek time.
+  //        6a2. Let animation's hold time be unresolved.
+  //        6a3. Apply any pending playback rate on animation.
+  if (seekTime != null) {
+    details.startTime = seekTime;
+    details.holdTime = null;
+    applyPendingPlaybackRate(details);
+  }
+
+  // Additional step for the polyfill.
+  addAnimation(details.timeline, details.animation,
+               tickAnimation.bind(details.proxy));
+
+  // 7. If animation's hold time is resolved, let its start time be
+  //    unresolved.
+  if (details.holdTime) {
+    details.startTime = null;
+  }
+
+  // 8. If animation has a pending play task or a pending pause task,
+  //   8.1 Cancel that task.
+  //   8.2 Set has pending ready promise to true.
+  if (details.readyPromise && details.proxy.pending) {
+    details.readyPromise.cancelTask();
+    hasPendingReadyPromise = true;
+  }
+
+  // 9. If the following three conditions are all satisfied:
+  //      animation’s hold time is unresolved, and
+  //      seek time is unresolved, and
+  //      aborted pause is false, and
+  //      animation does not have a pending playback rate,
+  //    abort this procedure.
+  if (details.holdTime === null && seekTime === null &&
+      !abortedPause && details.pendingPlaybackRate === null)
+  return;
+
+  // 10. If has pending ready promise is false, let animation’s current ready
+  //    promise be a new promise in the relevant Realm of animation.
+  if (details.readyPromise && !hasPendingReadyPromise)
+    details.readyPromise = null;
+
+  // Additional polyfill step to ensure that the native animation has the
+  // correct value for current time.
+  syncCurrentTime(details);
+
+  // 11. Schedule a task to run as soon as animation is ready.
+  if (!details.readyPromise)
+    createReadyPromise(details);
+  details.readyPromise.queueTask(commitPendingPlay, 'play');
```

Comment:
Why does this not use the promise infrastructure to run the task when the animation is ready? I.e. details.readyPromise.then(commitPendingPlay) or perhaps every time we create a ready promise we should details.readyPromise.then(commitPendingPlayState.bind(this)) and then that method can commit whatever is set as the pending play state?

Reply:
It is the responsibility of the pending play task to resolve the promise. Step 8 of:

https://www.w3.org/TR/web-animations-1/#playing-an-animation-section

We also need to be ale to cancel the pending task or to switch between a pending play and pending pause.

Reply:
createReadyPromise now calls notifyReady on the next animation frame or a call to tickAnimations provided that the timeline time is not null. NotifyReady will handle the task having been reset or switched since the time of promise creation.